src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -331,9 +331,28 @@ public IndexMetadata getIndexMetadata()
     public Callable<?> getInitializationTask()
     {
         // New storage-attached indexes will be available for queries after on disk index data are built.
-        // Memtable data will be indexed via flushing triggered by schema change
-        // We only want to validate the index files if we are starting up
-        IndexValidation validation = StorageService.instance.isStarting() ? IndexValidation.HEADER_FOOTER : IndexValidation.NONE;
+        // Memtable data will be indexed via flushing triggered by schema change.
+        // We only want to validate the index files if we are starting up.
+        boolean isStarting = StorageService.instance.isStarting();
+        IndexValidation validation = isStarting ? IndexValidation.HEADER_FOOTER : IndexValidation.NONE;
+
+        if (isStarting)
+        {
+            StorageAttachedIndexGroup indexGroup = StorageAttachedIndexGroup.getIndexGroup(baseCfs);
+            assert indexGroup != null : "Index group does not exist for table " + baseCfs.keyspace + '.' + baseCfs.name;
+
+            Collection<SSTableReader> nonIndexed = findNonIndexedSSTables(baseCfs, indexGroup, validation);
+
+            if (nonIndexed.isEmpty())
+            {
+                // If the index is complete, mark it queryable and avoid an initial build:
+                baseCfs.indexManager.makeIndexQueryable(this, Status.BUILD_SUCCEEDED);
+                logger.debug(indexIdentifier.logMessage("Skipping initial build, as index is already queryable..."));
+                initBuildStarted = true;
+                return () -> ImmediateFuture.success(null);
```

Comment:
Note that if we get here, the pre-join task will be a no-op. This actually resolves something that has been a problem for a while, which is that the async index build task and the pre-join task would simultaneously validate existing indexes on startup. We'll only do that work once now in the majority of cases.

Reply:
The one annoying case is when the index is actually damaged/incomplete on startup. In that case, we will actually validate twice, but that is because inside `startInitialBuild()`, we do a couple things to make sure the set of SSTables we need to index is correct (in the face of ongoing compactions, Memtable flushes, etc.)